---
layout: page
title: Science
permalink: /publications/
---

I'm a biological scientist with a strong interest in population and
evolutionary biology, particularly in the context of pathogens and
vectors of infectious diseases.

Below is a selection of scientific and technical articles that I had a
significant role in creating.

* Chris S. Clarkson, Alistair Miles et al. (2021) The genetic
  architecture of target‐site resistance to pyrethroid insecticides in
  the African malaria vectors *Anopheles gambiae* and *Anopheles
  coluzzii*. [https://doi.org/10.1111/mec.15845](https://doi.org/10.1111/mec.15845)

* The *Anopheles gambiae* 1000 Genomes Consortium (2020) Genome
  variation and population structure among 1142 mosquitoes of the
  African malaria vector species *Anopheles gambiae* and *Anopheles
  coluzzii*. [https://doi.org/10.1101/gr.262790.120](https://doi.org/10.1101/gr.262790.120)

* Eric R. Lucas, Alistair Miles et al. (2019) Whole-genome sequencing
  reveals high complexity of copy number variation at insecticide
  resistance loci in malaria
  mosquitoes. [https://doi.org/10.1101/gr.245795.118](https://doi.org/10.1101/gr.245795.118)

* The *Anopheles gambiae* 1000 Genomes Consortium (2017) Genetic
  diversity of the African malaria vector *Anopheles
  gambiae*. [https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6026373/](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6026373/)

* Alistair Miles et al. (2016) Indels, structural variation, and
  recombination drive genomic diversity in Plasmodium
  falciparum. [https://doi.org/10.1101/gr.203711.115](https://doi.org/10.1101/gr.203711.115)

* Alistair Miles et al. (2010) OpenFlyData: An exemplar data web
  integrating gene expression data on the fruit fly *Drosophila
  melanogaster*. [https://doi.org/10.1016/j.jbi.2010.04.004](https://doi.org/10.1016/j.jbi.2010.04.004)

* Alistair Miles and Sean Bechhofer (2009) SKOS Simple Knowledge
  Organization System
  Reference. [http://www.w3.org/TR/skos-reference](http://www.w3.org/TR/skos-reference)

* Diego Berrueta, Jon Phipps, Alistair Miles, Thomas Baker, Ralph
  Swick (2008) Best Practice Recipes for Publishing RDF
  Vocabularies. [http://www.w3.org/TR/swbp-vocab-pub/](http://www.w3.org/TR/swbp-vocab-pub/)

Since 2021 I have focussed on building data resources, software tools
and training resources to build capacity for other scientists to
interrogate and make discoveries from large-scale genomic data
resources. Here's a selection of papers and preprints that have built
on these efforts.

* McCann et al. (2024) Variation in spatial population structure in
  the Anopheles gambiae species
  complex. [https://doi.org/10.1101/2024.05.26.595955](https://doi.org/10.1101/2024.05.26.595955).

* Odero et al. (2024) Discovery of knock-down resistance in the major
  African malaria vector Anopheles
  funestus. [https://doi.org/10.1101/2024.03.13.584754](https://doi.org/10.1101/2024.03.13.584754)

* Kamau et al. (2024) The Anopheles coluzzii range extends into Kenya:
  detection, insecticide resistance profiles and population genetic
  structure in relation to conspecific populations in West and Central
  Africa. [https://doi.org/10.1186/s12936-024-04950-x](https://doi.org/10.1186/s12936-024-04950-x).

* della Torre et al. (2024) Speciation within the Anopheles gambiae
  complex: high-throughput whole genome sequencing reveals evidence of
  a putative new cryptic taxon in 'far-west'
  Africa. [https://doi.org/10.21203/rs.3.rs-3914444/v1](https://doi.org/10.21203/rs.3.rs-3914444/v1).

* Dennis et al. (2024) Cryptic population structure and insecticide
  resistance in Anopheles gambiae from the southern Democratic
  Republic of
  Congo. [https://doi.org/10.1101/2024.03.12.584605](https://doi.org/10.1101/2024.03.12.584605).

* Lucas et al. (2024) Copy number variants underlie the major
  selective sweeps in insecticide resistance genes in Anopheles
  arabiensis from
  Tanzania. [https://doi.org/10.1101/2024.03.11.583874](https://doi.org/10.1101/2024.03.11.583874).

* Nagi et al. (2024) Parallel evolution in mosquito vectors – a
  duplicated esterase locus is associated with resistance to
  pirimiphos-methyl in
  An. gambiae. [https://doi.org/10.1101/2024.02.01.578361](https://doi.org/10.1101/2024.02.01.578361).

* Kientega et al. (2023) Whole-genome sequencing of major malaria
  vectors revealed an evolution of new insecticide resistance variants
  in a longitudinal study in Burkina
  Faso. [https://doi.org/10.1101/2023.11.20.567800](https://doi.org/10.1101/2023.11.20.567800).

* Dyer et al. (2023) Mechanisms of transcriptional regulation in
  Anopheles gambiae revealed by allele specific
  expression. [https://doi.org/10.1101/2023.11.22.568226](https://doi.org/10.1101/2023.11.22.568226).

* Lucas et al. (2023) Genome-wide association studies reveal novel
  loci associated with pyrethroid and organophosphate resistance in
  Anopheles gambiae and Anopheles
  coluzzii. [https://doi.org/10.1038/s41467-023-40693-0](https://doi.org/10.1038/s41467-023-40693-0).

* Romero et al. (2023) Mitochondrial variation in Anopheles gambiae
  and An. coluzzii: phylogeographic legacy of species isolation and
  mito-nuclear associations with metabolic resistance to pathogens and
  insecticides. [https://doi.org/10.1101/2023.07.18.549472](https://doi.org/10.1101/2023.07.18.549472)

* St. Laurent et al. (2022) Population genomics reveal distinct and
  diverging populations of An. minimus in
  Cambodia. [https://doi.org/10.1038/s42003-022-04259-y](https://doi.org/10.1038/s42003-022-04259-y).

* Kientega et al. (2022) Analysis of the Genetic Variation of the
  Fruitless Gene within the Anopheles gambiae (Diptera: Culicidae)
  Complex Populations in
  Africa. [https://doi.org/10.3390/insects13111048](https://doi.org/10.3390/insects13111048).

* Wagah et al. (2021) Genetic variation at the Cyp6m2 putative
  insecticide resistance locus in Anopheles gambiae and Anopheles
  coluzzii. [https://doi.org/10.1186/s12936-021-03757-4](https://doi.org/10.1186/s12936-021-03757-4).

* Xue et al. (2021) Discovery of Ongoing Selective Sweeps within
  Anopheles Mosquito Populations Using Deep
  Learning. [https://doi.org/10.1093/molbev/msaa259](https://doi.org/10.1093/molbev/msaa259).